### PR TITLE
[SPARK-25425][SQL][BACKPORT-2.3] Extra options should override session options in DataSource V2

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

In the PR, I propose overriding session options by extra options in DataSource V2. Extra options are more specific and set via `.option()`, and should overwrite more generic session options. 

## How was this patch tested?

Added tests for read and write paths.

### CODE REVIEW

**Comment by MaxGekk** [comment]
@dongjoon-hyun Please, take a look at this PR. This is the backport of https://github.com/apache/spark/pull/22413 to Spark 2.3.

**Comment by MaxGekk** [comment]
@dongjoon-hyun @gatorsmile Can the fix be included into the upcoming minor release of 2.3?

**Comment by dongjoon-hyun** [comment]
I've considered this for 2.3.3 since 2.3.2 RC6 vote was already started. For now, I'm waiting the result of vote.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala [hunk 1]
```diff
@@ -315,19 +316,52 @@ class DataSourceV2Suite extends QueryTest with SharedSQLContext {
...
+          assert(!new File(sessionPath).exists)
+          checkAnswer(spark.read.format(format).option("path", optionPath).load(), df)
+        }
+      }
     }
   }
+}
 
-  override def createReader(options: DataSourceOptions): DataSourceReader = new Reader
+case class SimpleDataSourceV2Reader(options: DataSourceOptions) extends DataSourceReader {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2Suite.scala]: Thank you for this non-trivial backporting, @MaxGekk !

**Comment by dongjoon-hyun** [comment]
+1, LGTM. Merged to `branch-2.3`.

**Comment by dongjoon-hyun** [comment]
Thank you, @MaxGekk . Since this lands on `branch-2.3`, could you close this PR?

**Pull request closed without merging**